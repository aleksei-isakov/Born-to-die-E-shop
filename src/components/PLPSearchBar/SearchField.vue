<template>
  <div class="search-field">
    <input
      v-model="inputValue"
      class="search-field__input"
      :placeholder="placeHolder"
      @keypress.enter="onClickSearch"
    />
    <BaseButton class="search-field__btn" @click="onClickSearch">
      <BaseCustomIcon :icon="searchIcon" :width="searchIconWidth" />
    </BaseButton>
  </div>
</template>

<script>
import { BaseButton, BaseCustomIcon } from '@/base_components';

export default {
  name: 'SearchField',

  components: {
    BaseButton,
    BaseCustomIcon
  },

  props: {
    placeHolder: {
      type: String,
      default: '',
      required: true
    }
  },

  data() {
    return {
      searchIcon: 'search_icon',
      searchIconWidth: '50%',
      inputValue: ''
    };
  },

  methods: {
    onClickSearch() {
      this.$emit('search', this.inputValue.trim());
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.search-field {
  display: flex;
  justify-content: center;
  flex-grow: 1;
  height: 7vh;
  @media screen and (max-width: $mobile-size) {
    width: 100%;
  }
}

.search-field__input {
  width: 100%;
  border: 1px solid rgb(211, 210, 210);
  border-radius: 5px 0 0 5px;
  display: flex;
  justify-content: flex-end;
  padding-left: 10px;
}

.search-field__btn {
  width: 5vw;
  min-width: 50px;
  height: 100%;
  border: none;
  background-color: $primary;
  border-radius: 0 5px 5px 0;
  cursor: pointer;
}
</style>
