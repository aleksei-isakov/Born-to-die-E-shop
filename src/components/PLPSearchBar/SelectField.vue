<template>
  <div>
    <button
      class="select-field"
      @click="onClickToggleDropdown"
      @blur="onBlurCloseDropdown"
    >
      <div class="select-field__category">Category</div>
      <div>{{ currentCategory }}</div>
      <img
        class="select-field__arrow-icon"
        :src="arrowIcon"
        alt="arrow-icon"
        :class="{ reverse: isDropdownOpen }"
      />
    </button>
    <ul class="dropdown" :class="{ open: isDropdownOpen }">
      <select-field-item
        v-for="(category, i) in categories"
        :key="i"
        :category="category"
        :index="i"
        @chooseItem="onChooseItem"
      />
    </ul>
  </div>
</template>

<script>
import arrowIcon from '@/assets/Icons/arrow.svg';
import SelectFieldItem from './SelectFieldItem.vue';
import { mapActions } from 'vuex';

export default {
  name: 'SelectField',

  components: {
    SelectFieldItem
  },

  props: {
    categories: {
      type: Array,
      required: true,
      default: () => []
    }
  },

  data() {
    return {
      arrowIcon: arrowIcon,
      isDropdownOpen: false,
      activeCategoryIndex: 0
    };
  },

  computed: {
    currentCategory() {
      return this.categories[this.activeCategoryIndex];
    }
  },

  methods: {
    ...mapActions('PlpPageModule', ['setCurrentCategory']),

    onClickToggleDropdown() {
      this.isDropdownOpen = !this.isDropdownOpen;
    },

    onBlurCloseDropdown() {
      this.isDropdownOpen = false;
    },

    onChooseItem(index) {
      this.activeCategoryIndex = index;
      this.setCurrentCategory(this.currentCategory);
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.select-field {
  margin-right: 20px;
  width: 15vw;
  min-width: 120px;
  height: 7vh;
  background-color: $white;
  border: 1px solid rgb(211, 210, 210);
  border-radius: 5px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 10px;
  cursor: pointer;
}

.select-field__category {
  background-color: $white;
  color: #b3b4b6;
  padding: 0 5px 0 5px;
  position: absolute;
  top: -8px;
  left: 10px;
  font-size: 12px;
}

.dropdown {
  background-color: $white;
  border: 1px solid rgb(211, 210, 210);
  border-radius: 5px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 0px;
  width: 15vw;
  margin: 5px 0 0 0;
  opacity: 0;
  transition: $transition;
  height: 0px;
  min-width: 120px;
  max-height: 300px;
  overflow: auto;
  position: relative;
  z-index: 2;
}

@media screen and (max-width: $mobile-size) {
  .dropdown {
    width: 90vw;
    position: absolute;
  }
  .select-field {
    width: 90vw;
    margin: 0;
  }
}

.dropdown li {
  padding: 10px 10px;
  list-style: none;
  width: 100%;
  text-align: left;
  cursor: pointer;
}

.dropdown li:hover {
  background-color: rgb(238, 237, 237);
}

.select-field__arrow-icon {
  width: 10px;
  justify-self: right;
}

.open {
  height: fit-content;
  opacity: 1;
}

.reverse {
  transform: rotateX(180deg);
}
</style>
