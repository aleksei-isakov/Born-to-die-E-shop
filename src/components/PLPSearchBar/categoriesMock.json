["All Categories", "Category 1", "Category 2", "Category 3", "Category 4"]
