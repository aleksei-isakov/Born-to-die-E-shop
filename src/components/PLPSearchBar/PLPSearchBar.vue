<template>
  <div class="plp-search-bar">
    <select-field :categories="categories" />
    <search-field
      place-holder="Search by product name/description"
      @search="onSearch"
    />
  </div>
</template>

<script>
import SearchField from './SearchField.vue';
import SelectField from './SelectField.vue';
import { mapGetters } from 'vuex';

export default {
  name: 'PLPSearchBar',

  components: {
    SearchField,
    SelectField
  },

  computed: {
    ...mapGetters('PlpPageModule', ['categories'])
  },

  methods: {
    onSearch(inputValue) {
      this.$emit('search', inputValue);
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.plp-search-bar {
  width: 100%;
  display: flex;
  justify-content: space-between;
  gap: 20px;
  margin: 20px 0;
}

@media screen and (max-width: $mobile-size) {
  .plp-search-bar {
    flex-direction: column;
    align-items: center;
  }
}
</style>
