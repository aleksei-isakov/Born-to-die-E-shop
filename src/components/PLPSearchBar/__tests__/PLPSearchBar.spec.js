import { createLocalVue, shallowMount } from '@vue/test-utils';
import PLPSearchBar from '../PLPSearchBar.vue';
import Vuex from 'vuex';

const categoriesExample = ['a', 'b', 'c'];
let wrapper;
let getters;
let store;
const inputValue = 'my input';
const localVue = createLocalVue();

localVue.use(Vuex);

beforeEach(() => {
  getters = {
    categories: () => [categoriesExample]
  };

  store = new Vuex.Store({
    modules: {
      PlpPageModule: {
        namespaced: true,
        getters
      }
    }
  });

  wrapper = shallowMount(PLPSearchBar, {
    store,
    localVue
  });
});

afterEach(() => {
  wrapper.destroy();
});

describe('PLPSearchBar', () => {
  it('should emit search event', () => {
    wrapper.vm.onSearch(inputValue);
    expect(wrapper.emitted().search).toEqual([['my input']]);
  });

  it('should render correctly and match valid snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
