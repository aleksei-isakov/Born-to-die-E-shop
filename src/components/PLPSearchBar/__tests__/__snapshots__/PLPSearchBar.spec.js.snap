// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PLPSearchBar should render correctly and match valid snapshot 1`] = `
<div
  class="plp-search-bar"
>
  <select-field-stub
    categories="a,b,c"
  />
   
  <search-field-stub
    placeholder="Search by product name/description"
  />
</div>
`;
